src/app/model/ChessEngine.ts

```diff
@@ -67,4 +72,49 @@ export class ChessEngine implements ChessModel {
   private changeTurn(side: Side): Side {
     return side === Side.WHITE ? Side.BLACK : Side.WHITE;
   }
+
+  private willBeKingChecked(player: Player, chessboard: Chessboard, squareFrom: Square, squareTo: Square): boolean {
+    let isCheckedSquareFlag = false;
+
+    const proposedChessboard = _.cloneDeep(chessboard);
+    const { squaresWithPiece: proposedSquaresWithPieces } = proposedChessboard;
+
+    const movedPiece = chessboard.onPositionPiece(squareFrom);
+    const positionFrom = `${squareFrom.column}${squareFrom.row}`;
+    const positionTo = `${squareTo.column}${squareTo.row}`;
+    delete proposedSquaresWithPieces[positionFrom];
+    proposedSquaresWithPieces[positionTo] = movedPiece as Piece;
```

Comment:
Tutaj powtarzasz logikę z ChessBoard.
Fajnie byłoby zrobić jakąś metodę w stylu `chessBoardAfterMove(chessBoard: ChessBoard, /*to co trzeba*/): ChessBoard`, ktora zrobi Ci kopie, wykona niej ruch i zwroci nowa. Potem na tym mozesz sprawdzac czy krol jest szachowany.

Reply:
ok, ogarnę

Reply:
Dalej logika jest powtórzna. Wykorzystajmy to co mamy w Board :) to jest dokladnie to samo:

```.ts
  private returnWirtualProposedBoardAfterMove(chessboard: Chessboard, squareFrom: Square, squareTo: Square): Chessboard {
    const boardAfterMove = _.cloneDeep(chessboard);
    boardAfterMove.movePiece(squareFrom, squareTo);
    return boardAfterMove;
  }
```

Tez moze lepiej byloby to metode zrobic tak:
```.ts
  private boardAfterMove(chessboard: Chessboard, event: PieceWasMoved): Chessboard {
    const newBoard = _.cloneDeep(chessboard);
    newBoard.movePiece(event.from, event.to);
    return newBoard;
  }
```
A przynajmniej nazwe :) Bo nie wiem co znaczy 'Wirutal', nigdzie tez nie piszemy 'return'.

Reply:
Bez lodash kopia pewnie zadziala tak:
`const newBoard = new Chessboard({ ...chessboard.squaresWithPiece });`
Ale moze byc i lodash. Nie musimy robic deepCopy kazdej bierki z osobna, dlatego, ze one sa juz same w sobie immutable - nie maja zadnego stanu :) Dlatego jak jak z Buttonem pokazywalem - dobrze dazyc do takiego kodu. Jakbysmy np. robili tak tez ze kazdy ruch sachownicy nie modyfikuje tylko zwraca nowa szachownice to juz w ogole wtedy nie byloby potrzebne cos takiego jak teraz z kopiowaniem - ale juz nie poszlismy w ta strone :)

Reply:
Ok, Mateusz, kopiuję to z powodu tego, ze nie wykonuję tutaj uchu tylko go symuluję.
Poprawiłem nazwę. Wykorzystałem tak jak piszesz. Nie wiem czemu wcześniej na to nie wpadłem - zmyliło mnie wyrzucanie wyjątku i tylko skopiowałem. Będę pamiętał na przyszłość. 
Fajne rady